### PR TITLE
gn_build.sh: Add option to specify android sdk.

### LINKED ISSUE #5919: Configurable android SDK version in gn_build
#### Problem
android sdk version is hardcoded for gn_build (minor, but annoying)

 #### Proposed Solution
Add ability to change it

### PR DESCRIPTION
**Pull request by cecille**
Right now, this is hard coded in the config.

 #### Problem
android sdk version is hardcoded in the config

 #### Summary of Changes
Adds a check to gn_build.sh that allows this to be set as an env var along with the ANDROID_HOME / ANDROID_SDK_HOME

 Fixes #5919

### CODE REVIEW

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by mspang** [comment]
I don't think we should make every argument an environment variable. This one may be OK, but

1. You can pass arbitrary arguments like so:
  ```gn_build.sh android_sdk_version=<version>```

    Would that work well enough for you?

1. What's the reason you're building against a different SDK than the recommended one (which is the one our CI will test against) ?

**Review by jelderton — APPROVED** [review]

**Comment by cecille** [comment]
> I don't think we should make every argument an environment variable. This one may be OK, but
> 
> 1. You can pass arbitrary arguments like so:
>    `gn_build.sh android_sdk_version=<version>`
>    Would that work well enough for you?
> 2. What's the reason you're building against a different SDK than the recommended one (which is the one our CI will test against) ?

(sorry for the delay - I pushed this up right before I had a week vacation)

1) That's fine - I can change that if you'd like - I was just trying to match the style of the other android arguments
2) I'm actually OK to not put this in if you think it will cause problems, but basically I was 3 hours into trying to get stuff to build outside of the docker, I had just installed the wrong android SDK and NDK because I didn't know there was a specific version requirement and I didn't want to spend more time tracking down where whatever version was hosted. I you think it's important to compile against a specific version I'm OK to drop this CL, but we should update the documentation with a version number and a link to an official source for older versions of the android SDK.

**Comment by woody-apple** [comment]
@cecille Can you restyle this please? :)

**Comment by mspang** [comment]
> > I don't think we should make every argument an environment variable. This one may be OK, but
> > 
> > 1. You can pass arbitrary arguments like so:
> >    `gn_build.sh android_sdk_version=<version>`
> >    Would that work well enough for you?
> > 2. What's the reason you're building against a different SDK than the recommended one (which is the one our CI will test against) ?
> 
> (sorry for the delay - I pushed this up right before I had a week vacation)
> 
> 1. That's fine - I can change that if you'd like - I was just trying to match the style of the other android arguments

Not a style question, any arguments you add to the command line get passed through today. So:

`gn_build.sh android_sdk_version=21`

should work today and accomplish what your patch does with

`ANDROID_SDK_VERSION=21 gn_build.sh`

> 2. I'm actually OK to not put this in if you think it will cause problems, but basically I was 3 hours into trying to get stuff to build outside of the docker, I had just installed the wrong android SDK and NDK because I didn't know there was a specific version requirement and I didn't want to spend more time tracking down where whatever version was hosted. I you think it's important to compile against a specific version I'm OK to drop this CL, but we should update the documentation with a version number and a link to an official source for older versions of the android SDK.

I think it is a good idea to use the default. If you use a different version, your build will have different results (which may or may not be significant, but I'd recommend avoiding the risk especially if not working directly on the app). Note this script builds a bunch of things with the explicit goal of replicating GitHub CI and replicating CI means using the same inputs.

**Comment by cecille** [comment]
./gn_build.sh android_sdk_version=21 was my original attempt, but it throws an error for me.

ERROR at the command-line "--args":1:37: Build argument has no effect.
target_os="all" android_sdk_version=21
                                    ^-
The variable "android_sdk_version" was set as a build argument
but never appeared in a declare_args() block in any buildfile.

To view all possible args, run "gn args --list <out_dir>"


I get what you're saying about replicating what the CI, but when you're just getting up and running, there's nothing to tell you that you need version 21. Is there an official repo that houses that? The alternative here is to adjust the documentation so it's clear that the SDK has a required version and where to get it from. Otherwise I expect there are going to be a fair number of new CHIP-sters in this situation.

**Comment by mspang** [comment]
> ./gn_build.sh android_sdk_version=21 was my original attempt, but it throws an error for me.

Oops. https://github.com/project-chip/connectedhomeip/pull/6248

> 
> ERROR at the command-line "--args":1:37: Build argument has no effect.
> target_os="all" android_sdk_version=21
> ^-
> The variable "android_sdk_version" was set as a build argument
> but never appeared in a declare_args() block in any buildfile.
> 
> To view all possible args, run "gn args --list <out_dir>"
> 
> I get what you're saying about replicating what the CI, but when you're just getting up and running, there's nothing to tell you that you need version 21. Is there an official repo that houses that? The alternative here is to adjust the documentation so it's clear that the SDK has a required version and where to get it from. Otherwise I expect there are going to be a fair number of new CHIP-sters in this situation.

Yes, we should document it.

File: gn_build.sh [hunk 1]
```diff
@@ -110,10 +110,14 @@ android_sdk_args=""
...
 if [[ -d "${ANDROID_NDK_HOME}/toolchains" && -d "${ANDROID_HOME}/platforms" ]]; then
     android_sdk_args+="android_sdk_root=\"$ANDROID_HOME\" android_ndk_root=\"$ANDROID_NDK_HOME\""
+    if [[ ! -z "${ANDROID_SDK_VERSION}" ]]; then
+      android_sdk_args+="android_sdk_version=\"$ANDROID_SDK_VERSION\""
+    fi
     extra_args+=" $android_sdk_args enable_android_builds=true"
 else
     echo
     echo "Hint: Set \$ANDROID_HOME and \$ANDROID_NDK_HOME to enable building for Android"
+    echo "      Set \$CHIP_ANDROID_VERSION to change the android version (default 21)"
```
Comment (mspang) [inline gn_build.sh:120]: Let's turn this line into a note that Android SDK 21 is the one to use?

**Pull request closed without merging**